File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *reverse(ListNode *head){
        ListNode *prev = nullptr;
        ListNode *curr = head;
        while (curr){
            ListNode *temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
    int pairSum(ListNode* head) {
        ListNode *slow = head, *fast = head;

        while (fast!=nullptr && fast->next!=nullptr){
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode *rev = reverse(slow);
        int ans = INT_MIN;
        while (rev){
            int sum = rev->val + head->val;
            ans = max(ans, sum);
            rev = rev->next;
            head = head->next;
        }

        return ans;
    }
};

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reverse(node *ListNode) *ListNode {
    var prev *ListNode = nil
    curr := node
    for curr != nil {
        temp := curr.Next
        curr.Next = prev
        prev = curr
        curr = temp
    }

    return prev
}
func pairSum(head *ListNode) int {
    slow, fast := head, head
    for fast != nil && fast.Next != nil {
        slow = slow.Next
        fast = fast.Next.Next
    }
    left, right := head, reverse(slow)
    max_pair := 0
    for right != nil {
        pair := left.Val + right.Val
        max_pair = max(max_pair, pair)
        
        left = left.Next
        right = right.Next
    }

    return max_pair
}